File: backend/model/Livre.java
package com.bibliotheque.model;

// JPA : permet de lier cette classe à la table PostgreSQL
import jakarta.persistence.*;
// Lombok : génère automatiquement getters, setters, constructeurs
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;

import java.time.LocalDateTime;

// @Entity dit à Spring Boot :
// "Cette classe représente une table dans PostgreSQL"
@Entity

// @Table précise le nom exact de la table dans PostgreSQL
// C'est la table "livre" qu'on a créée dans pgAdmin
@Table(name = "livre")

// @Data (Lombok) génère automatiquement :
// - les getters (getId(), getTitre()...)
// - les setters (setId(), setTitre()...)
// - toString(), equals(), hashCode()
// Sans Lombok tu devrais écrire tout ça à la main !
@Data

// Génère un constructeur vide : new Livre()
// Obligatoire pour JPA
@NoArgsConstructor

// Génère un constructeur avec tous les champs :
// new Livre(titre, auteur, anneePublication, disponible)
@AllArgsConstructor

public class Livre {

    // @Id dit à JPA : "c'est la clé primaire de la table"
    @Id

    // @GeneratedValue dit à JPA : "laisse PostgreSQL
    // générer l'id automatiquement" (c'est notre BIGSERIAL)
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // @Column précise le nom de la colonne dans PostgreSQL
    // nullable = false → équivaut à NOT NULL dans SQL
    @Column(name = "titre", nullable = false)
    private String titre;

    @Column(name = "auteur", nullable = false)
    private String auteur;

    // Le champ anneePublication en Java
    // correspond à la colonne annee_publication en PostgreSQL
    @Column(name = "annee_publication", nullable = false)
    private Integer anneePublication;

    // Par défaut true : un livre créé est disponible
    // C'est notre DEFAULT TRUE dans PostgreSQL
    @Column(name = "disponible", nullable = false)
    private Boolean disponible = true;

    // Rempli automatiquement à la création
    // @Column(updatable = false) → cette valeur ne change jamais après création
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    // @PrePersist : cette méthode s'exécute automatiquement
    // juste AVANT que JPA insère l'enregistrement dans PostgreSQL
    // Tu n'as pas besoin de remplir createdAt manuellement
    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }
}

File: Frontend/models/livre.dart
class Livre {
  // Les mêmes champs que ta table PostgreSQL
  final int id;
  final String titre;
  final String auteur;
  final int anneePublication;
  final bool disponible;

  // Constructeur
  Livre({
    required this.id,
    required this.titre,
    required this.auteur,
    required this.anneePublication,
    required this.disponible,
  });

  // fromJson : convertit le JSON de Spring Boot
  // en objet Dart utilisable dans Flutter
  // Exemple JSON reçu :
  // {"id":1,"titre":"Clean Code","auteur":"Martin",...}
  factory Livre.fromJson(Map<String, dynamic> json) {
    return Livre(
      id: json['id'],
      titre: json['titre'],
      auteur: json['auteur'],
      // Spring Boot envoie "anneePublication"
      // exactement comme dans ton modèle Java
      anneePublication: json['anneePublication'],
      disponible: json['disponible'],
    );
  }

  // toJson : convertit l'objet Dart en JSON
  // pour envoyer à Spring Boot
  Map<String, dynamic> toJson() {
    return {
      'titre': titre,
      'auteur': auteur,
      'anneePublication': anneePublication,
      'disponible': disponible,
    };
  }
}